types/react/index.d.ts

```diff
@@ -82,7 +82,7 @@ declare namespace React {
 
     type JSXElementConstructor<P> =
         | ((props: P) => ReactElement | null)
-        | (new (props: P) => Component<P, any>);
+        | (new (props: P, context?: any) => Component<P, any>);
```

Comment:
It is not always receiving this additional parameter. 

> If contextTypes is defined within a component, the following lifecycle methods will receive an additional parameter, the context object:

Are you sure this additional parameter is safe or could users accidentally assume there's a second parameter?

Reply:
Yeah I'm a bit confused. The following code already works fine..
```ts
type ComponentProps<C> = C extends new (props: infer P) => React.Component<infer P> ? P : unknown

declare class MyComponent extends React.Component<{ foo: string }> {
    constructor(props: { foo: string }, context?: any)
 }

// $ExpectType { foo: string }
type Props = ComponentProps<typeof MyComponent>
```

Reply:
Yeah I was wondering why the [test doesn't pass on master](https://www.typescriptlang.org/play/?jsx=2&ssl=1&ssc=1&pln=9&pc=100#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoctAGxQGc64B1YGACwBkkBzdATwDCEAHYwkADxhDhdGFACuGaHAljhAE0bJ0MAHRDwIpKLgBvcnEtw0I2QqVQAFGBxg6ALjgphfADTWRMUlPbz4ASjMLK2i6eTAkJxcIN38bUVUwimiAX3Jc8hg+eOZWTh5+aSCpWzlFGGgABVdGAF5EYj0DSGFjGCbkugAeQviITBL2Ll40QUDVaTs66AA+UiA) but the problem is the test itself:
```diff
class WithLegacyContextConstructor extends React.Component {
-    constructor(props: any, context: any) {
+    constructor(props: any, context?: any) {
        super(props, context);
    }
}

```
fixes the test

Reply:
Which is the core issue here: With `contextTypes` the second argument isn't optional though we should probably allow it.

Reply:
Thanks for the comments, added `contextTypes` to the component in test to make it more realistic. 
Though I agree that it is weird to have context as a non-optional argument in the constructor, there might be a situation where we don't have access to the component code to fix it (e.g. third-party libraries), but still want to get target component props type.
It should be safe to make  JSXElementConstructor part that matches component constructor the same as  in other places like:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L493
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L586